### DISCUSSION TITLE
ComputeVariableEigenStrain

### DISCUSSION THREAD

**Discussion by uchiha-JN** [idaholab/moose#18797] [Q&A Modules: Solid mechanics] [answered]
I am using AD for my simulation and I need to use the material module "ComputeVariableEigenStrain" in TensorMechanics.
However, I cannot find its AD version and my input file won.t run otherwise . Can someone kindly help me out ?

Am I not searching in the right place or does it not exist?

**Comment by GiudGiud** [comment, marked as answer]
@hugary1995

**Reply by jiangwen84** [reply]
I guess it does not exist. You should be able to mix AD with non-AD though.

**Reply by uchiha-JN** [reply]
Okay... Thanks for your reply.  
1. I did not know AD and non-AD could be mixed . Thanks for the info. It gives me hope and an idea
2. I keep getting errors about some properties being declared as AD while being requested as regular and vice versa. I guess that will require me to get into the CPP codes and header files. an example of the error message is below

error message
*** ERROR ***
The requested AD material property mechanical_strain is declared as a regular material property. Either retrieve it as a regular material property with getMaterialProperty or declare it as an AD property with declareADProperty.

**Reply by GiudGiud** [reply]
It can be mixed though there might be some difficulties on the way. 
For example for this property (which is defined behind the scenes by the Tensor_Mechanics action I think), if you cannot sub the non_AD kernel that requires it with an AD version, then you will have to make a non-AD copy of the material property, will obviously loose the derivatives information.

**Reply by uchiha-JN** [reply]
Gotcha... It did come with difficulties. Just like you said. And yes, I believe you are right about me losing the derivative information. 
However, I managed to make a copy like you suggested and changed a few things to make it work just fine. 
Thank you @GiudGiud  and @jiangwen84  for your help. I really appreciate it.
I am now running a simulation with the new kernel and I hope I do not run into any more issues :)

**Reply by hugary1995** [reply]
Sorry I'm late on this. Glad you've figured it out! You can also make a pull request to add the AD version of CompiteVariableEigenstrain to moose. I believe such contribution is very welcome.
